<a href="https://colab.research.google.com/github/purnima0122/Flight-Delay-Data-Pipeline/blob/main/Airline_delay.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Airline Delay Cause — Data Pipeline & Visualization

This dataset contains monthly flight arrival and delay data for U.S. airports, categorized by airline carriers. It includes total arriving flights, delays over 15 minutes, cancellations, diversions, and delay causes such as carrier issues, weather, NAS, security, and late aircraft arrivals.

### **Importing Necessary Libraries**

In [3]:
#importing the basic libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### **Step 1: EXTRACT**

### Importing the dataset

In [4]:
flight_delay = pd.read_csv('Airline_Delay_Cause.csv')  # load CSV from project folder


### Analyzing the dataset

 Let's take a peek into first few rows of the iris dataset

In [5]:
flight_delay.head()# displaying the first 5 rows of the dataset

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,security_ct,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
0,2023,8,9E,Endeavor Air Inc.,ABE,"Allentown/Bethlehem/Easton, PA: Lehigh Valley ...",89.0,13.0,2.25,1.60,...,0.0,5.99,2.0,1.0,1375.0,71.0,761.0,118.0,0.0,425.0
1,2023,8,9E,Endeavor Air Inc.,ABY,"Albany, GA: Southwest Georgia Regional",62.0,10.0,1.97,0.04,...,0.0,7.42,0.0,1.0,799.0,218.0,1.0,62.0,0.0,518.0
2,2023,8,9E,Endeavor Air Inc.,AEX,"Alexandria, LA: Alexandria International",62.0,10.0,2.73,1.18,...,0.0,4.28,1.0,0.0,766.0,56.0,188.0,78.0,0.0,444.0
3,2023,8,9E,Endeavor Air Inc.,AGS,"Augusta, GA: Augusta Regional at Bush Field",66.0,12.0,3.69,2.27,...,0.0,1.57,1.0,1.0,1397.0,471.0,320.0,388.0,0.0,218.0
4,2023,8,9E,Endeavor Air Inc.,ALB,"Albany, NY: Albany International",92.0,22.0,7.76,0.00,...,0.0,11.28,2.0,0.0,1530.0,628.0,0.0,134.0,0.0,768.0


In [6]:
flight_delay.tail()# displaying the last 5 rows of the dataset

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,security_ct,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
171661,2013,8,YV,Mesa Airlines Inc.,TRI,"Bristol/Johnson City/Kingsport, TN: Tri Cities",7.0,1.0,0.78,0.0,...,0.0,0.0,0.0,0.0,23.0,18.0,0.0,5.0,0.0,0.0
171662,2013,8,YV,Mesa Airlines Inc.,YUM,"Yuma, AZ: Yuma MCAS/Yuma International",2.0,1.0,0.00,0.0,...,0.0,1.0,0.0,0.0,17.0,0.0,0.0,0.0,0.0,17.0
171663,2013,8,YV,Mesa Airlines Inc.,SLC,"Salt Lake City, UT: Salt Lake City International",2.0,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
171664,2013,8,DL,Delta Air Lines Inc.,SBN,"South Bend, IN: South Bend International",1.0,1.0,1.00,0.0,...,0.0,0.0,0.0,0.0,18.0,18.0,0.0,0.0,0.0,0.0
171665,2013,8,F9,Frontier Airlines Inc.,MTJ,"Montrose/Delta, CO: Montrose Regional",1.0,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


### Statistical Insights

 Let's see quick overview of the basic statistics of the data without having to calculate each of these things individually.

In [7]:
display(flight_delay.describe().T)# This gives you the basic statistical values just for numeric columns

,count,mean,std,min,25%,50%,75%,max
year,171666.0,2018.551361,2.890006,2013.0,2016.00,2019.00,2021.00,2023.00
month,171666.0,6.493633,3.440908,1.0,4.00,7.00,9.00,12.00
arr_flights,171426.0,362.528467,992.894662,1.0,50.00,100.00,250.00,21977.00
arr_del15,171223.0,66.434387,179.540694,0.0,6.00,17.00,47.00,4176.00
carrier_ct,171426.0,20.796615,50.315176,0.0,2.16,6.40,17.26,1293.91
weather_ct,171426.0,2.250347,7.314252,0.0,0.00,0.40,1.86,266.42
nas_ct,171426.0,19.381147,61.675244,0.0,1.00,3.91,11.71,1884.42
security_ct,171426.0,0.157096,0.717405,0.0,0.00,0.00,0.00,58.69
late_aircraft_ct,171426.0,23.770554,72.393477,0.0,1.23,5.00,15.26,2069.07
arr_cancelled,171426.0,7.530497,43.654880,0.0,0.00,1.00,4.00,4951.00


Now, let's take a quick peek to see the size of the data we are working with.

In [8]:
flight_delay.shape #rows vs columns

(171666, 21)

We have 171666 rows and 21 columns

Let's see the summary of dataset with columns its data types and number of non null values

In [9]:
flight_delay.info() #summary of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 171666 entries, 0 to 171665
Data columns (total 21 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   year                 171666 non-null  int64  
 1   month                171666 non-null  int64  
 2   carrier              171666 non-null  object 
 3   carrier_name         171666 non-null  object 
 4   airport              171666 non-null  object 
 5   airport_name         171666 non-null  object 
 6   arr_flights          171426 non-null  float64
 7   arr_del15            171223 non-null  float64
 8   carrier_ct           171426 non-null  float64
 9   weather_ct           171426 non-null  float64
 10  nas_ct               171426 non-null  float64
 11  security_ct          171426 non-null  float64
 12  late_aircraft_ct     171426 non-null  float64
 13  arr_cancelled        171426 non-null  float64
 14  arr_diverted         171426 non-null  float64
 15  arr_delay        

* We see all the columns of our data set are free from null values and four of them are numeric while one is a categorical column which also is seen by the dtypes function.

In [10]:
flight_delay.dtypes #data types of each column

,0
year,int64
month,int64
carrier,object
carrier_name,object
airport,object
airport_name,object
arr_flights,float64
arr_del15,float64
carrier_ct,float64
weather_ct,float64


In [11]:
flight_delay.isnull().sum() #Missing value counts

,0
year,0
month,0
carrier,0
carrier_name,0
airport,0
airport_name,0
arr_flights,240
arr_del15,443
carrier_ct,240
weather_ct,240


In [12]:
flight_delay.duplicated().sum()#check for duplicates

np.int64(0)

Transform: Data Cleaning

In [13]:
flight_delay=flight_delay.drop_duplicates()

In [22]:
flight_delay["calculated_delay"] = (
    flight_delay["carrier_delay"] +
    flight_delay["weather_delay"] +
    flight_delay["nas_delay"] +
    flight_delay["security_delay"] +
    flight_delay["late_aircraft_delay"]
)

(flight_delay["arr_delay"] - flight_delay["calculated_delay"]).describe()

,0
count,171426.000000
mean,0.004218
std,1.316511
min,0.000000
25%,0.000000
50%,0.000000
75%,0.000000
max,509.000000


We validated whether total arrival delay equals the sum of delay causes.
75% of the records matched exactly.
The mean difference was 0.004 minutes, indicating high data integrity.
A few extreme outliers were detected (max difference = 509), possibly due to rounding or reporting inconsistencies.
Overall, the dataset shows strong internal consistency.

### **Step 2: TRANSFORM**

The transformation phase covers:
1. **Drop/fix data types** — ensure numeric columns are truly numeric
2. **Handle missing values** — fill or drop NaNs with justification
3. **Remove duplicates**
4. **Outlier awareness** — log note; delay data is naturally right-skewed
5. **Feature engineering** — derive useful new columns for analysis

In [15]:
flight_delay=flight_delay.drop_duplicates()

In [16]:
flight_delay[
    (flight_delay["arr_delay"] - flight_delay["calculated_delay"]).abs() > 10
]

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay,calculated_delay
48527,2021,3,F9,Frontier Airlines Inc.,STL,"St. Louis, MO: St Louis Lambert International",87.0,17.0,2.59,0.00,...,4.87,3.0,0.0,559.0,99.0,0.0,261.0,0.0,179.0,539.0
59109,2020,8,DL,Delta Air Lines Inc.,SEA,"Seattle, WA: Seattle/Tacoma International",1842.0,127.0,52.19,6.54,...,28.48,6.0,1.0,7690.0,4059.0,269.0,1203.0,21.0,1629.0,7181.0
117233,2017,8,F9,Frontier Airlines Inc.,TTN,"Trenton, NJ: Trenton Mercer",187.0,44.0,6.48,0.06,...,14.56,5.0,2.0,4044.0,682.0,6.0,1401.0,0.0,1761.0,3850.0


###Create Meaningful Features

####Delay Rate (%)

In [17]:
flight_delay["delay_rate"] = (
    flight_delay["arr_del15"] / flight_delay["arr_flights"]
) * 100

This normalizes delays by total flights.

####Cancellation Rate (%)

In [18]:
flight_delay["cancel_rate"] = (
    flight_delay["arr_cancelled"] / flight_delay["arr_flights"]
) * 100

####Diversion Rate (%)

In [19]:
flight_delay["divert_rate"] = (
    flight_delay["arr_diverted"] / flight_delay["arr_flights"]
) * 100

####Average Delay per Flight

In [20]:
flight_delay["avg_delay_per_flight"] = (
    flight_delay["arr_delay"] / flight_delay["arr_flights"]
)

Airports with more flights naturally have larger total delays

In [21]:
flight_delay.head()

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay,calculated_delay,delay_rate,cancel_rate,divert_rate,avg_delay_per_flight
0,2023,8,9E,Endeavor Air Inc.,ABE,"Allentown/Bethlehem/Easton, PA: Lehigh Valley ...",89.0,13.0,2.25,1.60,...,71.0,761.0,118.0,0.0,425.0,1375.0,14.606742,2.247191,1.123596,15.449438
1,2023,8,9E,Endeavor Air Inc.,ABY,"Albany, GA: Southwest Georgia Regional",62.0,10.0,1.97,0.04,...,218.0,1.0,62.0,0.0,518.0,799.0,16.129032,0.000000,1.612903,12.887097
2,2023,8,9E,Endeavor Air Inc.,AEX,"Alexandria, LA: Alexandria International",62.0,10.0,2.73,1.18,...,56.0,188.0,78.0,0.0,444.0,766.0,16.129032,1.612903,0.000000,12.354839
3,2023,8,9E,Endeavor Air Inc.,AGS,"Augusta, GA: Augusta Regional at Bush Field",66.0,12.0,3.69,2.27,...,471.0,320.0,388.0,0.0,218.0,1397.0,18.181818,1.515152,1.515152,21.166667
4,2023,8,9E,Endeavor Air Inc.,ALB,"Albany, NY: Albany International",92.0,22.0,7.76,0.00,...,628.0,0.0,134.0,0.0,768.0,1530.0,23.913043,2.173913,0.000000,16.630435


## **Data Visualization**